arangod/Replication/DatabaseInitialSyncer.cpp

```diff
@@ -968,111 +972,141 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
+        return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
+      }
+
+      std::chrono::milliseconds sleepTime = ::sleepTimeFromWaitTime(waitTime);
+      std::this_thread::sleep_for(sleepTime);
     }
 
-    if (static_cast<uint64_t>(waitTime * 1000.0 * 1000.0) >= _config.applier._initialSyncMaxWaitTime) {
+    stats.waitedForInitial += TRI_microtime() - startTime;
+
+    if (replutils::hasFailed(response.get())) {
       ++stats.numFailedConnects;
-      stats.waitedForInitial += waitTime;
-      return Result(TRI_ERROR_REPLICATION_NO_RESPONSE,
-                    std::string(
-                        "timed out waiting for response from leader at ") +
-                        _config.leader.endpoint);
+      return replutils::buildHttpError(response.get(), url, _config.connection);
     }
 
-    if (isAborted()) {
-      stats.waitedForInitial += waitTime;
-      return Result(TRI_ERROR_REPLICATION_APPLIER_STOPPED);
+    Result r = replutils::parseResponse(builder, response.get());
+
+    if (r.fail()) {
+      ++stats.numFailedConnects;
+      return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+                    std::string("got invalid response from leader at ") +
+                    _config.leader.endpoint + url + ": " + r.errorMessage());
     }
 
-    std::chrono::milliseconds sleepTime = ::sleepTimeFromWaitTime(waitTime);
-    std::this_thread::sleep_for(sleepTime);
-  }
-  
-  stats.waitedForInitial += TRI_microtime() - startTime;
+    slice = builder.slice();
+    if (!slice.isObject()) {
+      ++stats.numFailedConnects;
+      return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
+                    std::string("got invalid response from leader at ") +
+                    _config.leader.endpoint + url + ": response is no object");
+    }
 
-  if (replutils::hasFailed(response.get())) {
-    ++stats.numFailedConnects;
-    return replutils::buildHttpError(response.get(), url, _config.connection);
-  }
+    return Result();
 
-  VPackBuilder builder;
-  Result r = replutils::parseResponse(builder, response.get());
+  };
 
-  if (r.fail()) {
-    ++stats.numFailedConnects;
-    return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                  std::string("got invalid response from leader at ") +
-                      _config.leader.endpoint + url + ": " + r.errorMessage());
-  }
 
-  VPackSlice const slice = builder.slice();
-  if (!slice.isObject()) {
-    ++stats.numFailedConnects;
-    return Result(TRI_ERROR_REPLICATION_INVALID_RESPONSE,
-                  std::string("got invalid response from leader at ") +
-                      _config.leader.endpoint + url + ": response is no object");
+  auto ck = keysCall(true);
+  if (!ck.ok()) {
+    return ck;
+  }
+  if (!slice.hasKey("id")) { // we only have count
+    VPackSlice const c = slice.get("count");
+    if (c.isNumber()) {
+      maxWaitTime = c.getNumber<uint64_t>() * 864/10000000; // ~ pessimistic 1 billion / day
```

Comment:
I can't follow this calculation.
It uses integer arithmetic (no floating point values used). So assumed we get a count value of 0 back, then the calculated maxWaitTime is 0, too.
If we get a count value of 10k back, the calculated maxWaitTime will be 0.
If we get a count value of 1M back, the calculated maxWaitTime will be 86.
If we get a count value of 100M back, the calculated maxWaitTime will be 8640.

Assumed that maxWaitTime should have a unit of microseconds (https://github.com/arangodb/arangodb/pull/13394/files?diff=unified&w=1#diff-db234bdc91c0f395720e4bf5d69da8fbc541f5d66a6f73aac78ab6df02b7a387R1046), I think these calculated values are not sensible.

Reply:
In addition, we should not go below the already configured `maxWaitTime` value here.